File: wallendar-app/app/src/main/java/com/td/wallendar/login/ui/LoginPresenter.java
package com.td.wallendar.login.ui;

import com.td.wallendar.dtos.request.LoginRequest;
import com.td.wallendar.dtos.response.LoginResponse;
import com.td.wallendar.models.ApplicationUser;
import com.td.wallendar.repositories.interfaces.ApplicationUsersRepository;
import com.td.wallendar.utils.scheduler.SchedulerProvider;

import java.lang.ref.WeakReference;

import io.reactivex.disposables.CompositeDisposable;

public class LoginPresenter {
    private final WeakReference<LoginView> loginView;
    private final ApplicationUsersRepository applicationUsersRepository;
    private final CompositeDisposable disposable;
    private final SchedulerProvider schedulerProvider;

    public LoginPresenter(LoginView loginView, ApplicationUsersRepository applicationUsersRepository, SchedulerProvider schedulerProvider) {
        this.loginView = new WeakReference<>(loginView);
        this.applicationUsersRepository = applicationUsersRepository;
        this.schedulerProvider = schedulerProvider;
        this.disposable = new CompositeDisposable();
    }

    public void attemptLogin(final String email, final String password) {
        disposable.add(applicationUsersRepository.authenticateUser(new LoginRequest(email, password))
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe(this::onLoginAttemptSuccessful, this::onLoginAttemptError));
    }

    private void onLoginAttemptSuccessful(LoginResponse loginResponse) {
        getApplicationUser(loginResponse.getEmail());
    }

    private void onLoginAttemptError(Throwable throwable) {
        loginView.get().loginFailed();
    }

    private void getApplicationUser(final String email) {
        disposable.add(applicationUsersRepository.getUserByEmail(email)
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe(this::onFindUserSuccessful, this::onFindUserFailed));
    }

    private void onFindUserSuccessful(ApplicationUser applicationUser) {
        loginView.get().loginSuccessful(applicationUser.getId());
    }

    private void onFindUserFailed(Throwable throwable) {
        loginView.get().loginFailed();
    }

    public void onViewDetached() {
        disposable.dispose();
    }
}

File: wallendar-app/app/src/main/java/com/td/wallendar/login/ui/LoginPresenter.kt
package com.td.wallendar.login.ui

import com.td.wallendar.dtos.request.LoginRequest
import com.td.wallendar.dtos.response.LoginResponse
import com.td.wallendar.models.ApplicationUser
import com.td.wallendar.repositories.interfaces.ApplicationUsersRepository
import com.td.wallendar.utils.scheduler.SchedulerProvider
import io.reactivex.disposables.CompositeDisposable
import java.lang.ref.WeakReference

class LoginPresenter(loginView: LoginView, private val applicationUsersRepository: ApplicationUsersRepository, private val schedulerProvider: SchedulerProvider) {
    private val loginView: WeakReference<LoginView> = WeakReference(loginView)
    private val disposable: CompositeDisposable = CompositeDisposable()

    fun attemptLogin(email: String, password: String) {
        disposable.add(applicationUsersRepository.authenticateUser(LoginRequest(email, password))
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe({ loginResponse: LoginResponse -> onLoginAttemptSuccessful(loginResponse) }) { throwable: Throwable -> onLoginAttemptError(throwable) })
    }

    private fun onLoginAttemptSuccessful(loginResponse: LoginResponse) {
        getApplicationUser(loginResponse.email)
    }

    private fun onLoginAttemptError(throwable: Throwable) {
        loginView.get()?.loginFailed()
    }

    private fun getApplicationUser(email: String) {
        disposable.add(applicationUsersRepository.getUserByEmail(email)
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe({ applicationUser: ApplicationUser -> onFindUserSuccessful(applicationUser) }) { throwable: Throwable -> onFindUserFailed(throwable) })
    }

    private fun onFindUserSuccessful(applicationUser: ApplicationUser) {
        loginView.get()?.loginSuccessful(applicationUser.id)
    }

    private fun onFindUserFailed(throwable: Throwable) {
        loginView.get()?.loginFailed()
    }

    fun onViewDetached() {
        disposable.dispose()
    }

}